Respect per-shard tablet goal and 10x default per-shard tablet count

tgrabiec opened the pull request:
This series achieves two things:

1) changes default number of tablet replicas per shard to be 10 in order to reduce load imbalance between shards

    This will result in new tables having at least 10 tablet replicas per
    shard by default.
    
    We want this to reduce tablet load imbalance due to differences in
    tablet count per shard, where some shards have 1 tablet and some
    shards have 2 tablets. With higher tablet count per shard, this
    difference-by-one is less relevant.
  
    Fixes https://github.com/scylladb/scylladb/issues/21967

2) introduces a global goal for tablet replica count per shard and adds logic to tablet scheduler to respect it by controlling per-table tablet count 
    
    The per-shard goal is enforced by controlling average per-shard tablet replica
    count in a given DC, which is controlled by per-table tablet
    count. This is effective in respecting the limit on individual shards
    as long as tablet replicas are distributed evenly between shards.
    There is no attempt to move tablets around in order to enforce limits
    on individual shards in case of imbalance between shards.
    
    If the average per-shard tablet count exceeds the limit, all tables
    which contribute to it (have replicas in the DC) are scaled down
    by the same factor. Due to rounding up to the nearest power of 2,
    we may overshoot the per-shard goal by at most a factor of 2.

    The scaling is applied after computing desired tablet count due to
    all other factors: per-table tablet count hints, defaults, average tablet size. 
    
    If different DCs want different scale factors of a given table, the
    lowest scale factor is chosen for a given table.

    When creating a new table, its tablet count is determined by tablet
    scheduler using the scheduler logic, as if the table was already created.
    So any scaling due to per-shard tablet count goal is reflected immediately
    when creating a table. It may however still take some time for the system
    to shrink existing tables. We don't reject requests to create new tables.  

    Fixes #21458

scylladb-promoter commented:
## Docs Preview :book:
Docs Preview for this pull request is available [here](http://previews.docs.scylladb.com/scylladb/22522)

**Changed Files:**
- [cql/ddl](http://previews.docs.scylladb.com/scylladb/22522/cql/ddl)

**Note:** This preview will be available for 30 days and will be automatically deleted after that period. You can manually trigger a new build by committing changes.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14776/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14776/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/network_topology_strategy_test
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_major_compaction
:small_blue_diamond: topology_custom/test_tablets
#### Failed Tests (281/1342):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.1.xunit/xml/_empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.1.xunit/xml/_empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.10.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.10.xunit/xml/_empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.10.xunit/xml/_empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.2.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.2.xunit/xml/_empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.2.xunit/xml/_empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/14776/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.3.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 2 hr 51 min
- Builder: spider8.cloudius-systems.com

mykaul commented:
@avikivity -  please review.

tgrabiec commented:
Review ping @avikivity

avikivity commented on locator/network_topology_strategy.cc:
```diff
@@ -311,7 +311,7 @@ static future<unsigned> calculate_initial_tablets_from_topology(const schema& s,
     std::unordered_map<sstring, unsigned> shards_per_dc_map;
     tm->get_topology().for_each_node([&] (const node& n) {
-        if (tm->is_normal_token_owner(n.host_id())) {
+        if (n.is_normal() && tm->is_normal_token_owner(n.host_id())) {
```
Shouldn't this be taken out and backported?

tgrabiec replied:
Yes, done: https://github.com/scylladb/scylladb/pull/22657

avikivity commented on locator/network_topology_strategy.cc:
```diff
@@ -315,16 +315,13 @@ static future<unsigned> calculate_initial_tablets_from_topology(const schema& s,
-            auto min_tablets_in_dc = std::ceil((double)(min_per_shard_tablet_count * shards_in_dc) / rf_in_dc);
-            tablets_in_dc = std::max<unsigned>(tablets_in_dc, min_tablets_in_dc);
-        }
+        unsigned tablets_in_dc = std::ceil((double)(min_per_shard_tablet_count * shards_in_dc) / rf_in_dc);
```
I understand it's just a simplification, but I would like to allow values 0 < min_per_shard_tablet_count < 1 one day. For example 0.5 would mean that we target half the shards in the cluster to have a tablet. But not worth dwelling on now.

bhalevy replied:
Agreed, in large clusters serving a large number of small tables, shards will have to serve a subset of the tables to satisfy the global per-shard tablet count goal.

tgrabiec replied:
I agree too.

bhalevy replied:
The question is whether we need to keep the current logic.
I think that if the `min_per_shard_tablet_count` option is specified and is less than one we should respect it.
If it isn't specified than the previous patch will use the initial scale factor or 1 which satisfies the max expression above.

tgrabiec replied:
In V3 I changed the code to respect min_per_shard_tablet_count / initial_scale smaller than 1.

avikivity commented on locator/network_topology_strategy.cc:
```diff
@@ -350,14 +350,7 @@ future<size_t> network_topology_strategy::calculate_min_tablet_count(schema_ptr
     co_return tablet_count;
 }
 
```
I don't understand the goal.

tgrabiec replied:
Is this comment on the "tablets: Use scheduler's make_sizing_plan() to decide about tablet count of a new table" patch? (https://github.com/scylladb/scylladb/pull/22522/commits/434f9150e41b741caacb25336a2e2bd9ed2e0540) github detaches comment from commit, so I'm not sure.

The goal is explained in the previous commit, "tablets: load_balancer: Extract make_sizing_plan()":


> Resize plan making will now happen in two stages:
> 
>   1) Determine desired tablet counts per table (sizing plan)
>   2) Schedule resize decisions
> 
> We need intermediate step in the resize plan making, which gives us
> the planned tablet counts, so that we can plug this part of the
> algorithm into initial tablet allocation on table construction.
> 
> We want decisisons made by the scheduler to be consistent with
> decisions made on table creation. We want to avoid over-allocation of
> tablets when table is created, which would then be reduced by the
> scheduler. Not just to avoid wasteful migrations post table creation,
> but to respect the per-shard goal. To respect the per-shard goal, the
> algorithm will no longer be as simple as looking at hints, and we want
> to share the algorithm between the scheduler and initial tablet
> allocator.
> 
> Also, this sizing plan will be later plugged into a virtual table for
> observability.
>

tgrabiec replied:
Ping @avikivity

avikivity replied:
I'm sure it was on a different patch. I'll look for it.

avikivity replied:
I can't find it. Maybe the patch was removed or changed?

tgrabiec replied:
I didn't push to the branch since the PR was opened. But the comment is marked "Outdated". I don't know how to explain it.

avikivity commented on locator/tablet_replication_strategy.hh:
```diff
@@ -50,6 +50,12 @@ public:
+    /// Returns desired replication factor in a given DC.
+    /// Note that individual tablets may lag behind desired replication factor in their
+    /// current replica list, as replication factor changes invole table rebuilding transitions
+    /// which are not instantenous.
```
instantaneous, don't get @tchaikov's spelling bot angry at you

bhalevy replied:
If it is the desired replication factor rather than the actual one, I suggest to reflect it in the method name, e.g. `get_desired_replication_factor`.

tgrabiec replied:
But there's no other kind of replication factor, so it's redundant to add extra qualifiers. 

This doesn't add a new method, only lifts it to the interface.

bhalevy replied:
Maybe drop the word "desired" from the comment then.
It's confusing IMO.

tgrabiec replied:
Ok, I can drop it.

bhalevy replied:
Still not dropped in latest version.
Also, there's a typo in the comment: s/invole/involve/

avikivity commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+        // by a factor of 2 in the worst case. If we choose a subset of tables to scale down by a factor of 2 then
+        // we have a problem of making sure that the choice is stable across scheduler invocations to avoid
+        // oscillations of decisions.
+
```
Agree with this long comment

avikivity commented:
Looks reasonable.

bhalevy commented:
@tgrabiec #22090 was merged, please rebase

bhalevy commented on db/config.cc:
```diff
@@ -1325,7 +1325,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , maximum_replication_factor_fail_threshold(this, "maximum_replication_factor_fail_threshold", liveness::LiveUpdate, value_status::Used, -1, "")
-    , tablets_initial_scale_factor(this, "tablets_initial_scale_factor", value_status::Used, 10, "Calculated initial tablets are multiplied by this number")
+    , tablets_initial_scale_factor(this, "tablets_initial_scale_factor", value_status::Used, 10,
+         "Minimum average number of tablet replicas per shard per table. Suppressed by tablet hints in table's schema: min_per_shard_tablet_count and min_tablet_count")
```
nit: rename `hints` to `options`
Also, how about the expected_table_size_in_gb option?  Maybe we should list the particular option and leave this open to any per-table option may take priority over the configuration option, which can be described as just the default.

bhalevy replied:
nit in commit `tablets, topology: Introduce tablets_per_shard_goal config` log message: s/decrese/decrease/

tgrabiec replied:
> nit: rename `hints` to `options` 

Yes

> Also, how about the expected_table_size_in_gb option? Maybe we should list the particular option and leave this open to any per-table option may take priority over the configuration option, which can be described as just the default.

expected_table_size_in_gb doesn't suppress it, because it doesn't lower the minimum. It would suppress it if the count could go below that which follows from `tablets_initial_scale_factor` logic. But the meaning of the option is to set a minimum, not a particular count, so if the count goes higher, the option is still in effect.

bhalevy commented on locator/network_topology_strategy.cc:
```diff
@@ -340,12 +340,16 @@ future<size_t> network_topology_strategy::calculate_min_tablet_count(schema_ptr
+    // This is for compatibility with the old "initial" tablet count option which suppresses automatic scaling.
+    auto min_per_shard_tablet_count = tablet_hints.min_per_shard_tablet_count.value_or(
+            tablet_hints.min_tablet_count ? 0 : initial_scale.value_or(1));
+    if (min_per_shard_tablet_count) {
```
Don't we need to call `calculate_initial_tablets_from_topology` if `tablet_count == 0`, even if the min per-shard count is zero?

tgrabiec replied:
hmm, that depends on whether `min_per_shard_tablet_count == 0` means that the option is unset (then your suggestion applies), or it means that we want to force no minimum, and go by avg tablet size alone. I assumed the latter.

tgrabiec replied:
If someone wants to unset the option, he can remove it from tablet options.

bhalevy replied:
let's look at this again after rebase.

bhalevy replied:
My concern was about the default case for new tables where no option helped determine the tablet_count, so we defer to `calculate_initial_tablets_from_topology`.
The average tablet size for new tables would be 0.

tgrabiec replied:
If there are no tablet options, then `min_per_shard_tablet_count` will pick up the default `initial_scale` option.

bhalevy replied:
ok, got it.

bhalevy commented on db/config.cc:
```diff
@@ -1327,6 +1327,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , tablets_initial_scale_factor(this, "tablets_initial_scale_factor", value_status::Used, 10,
          "Minimum average number of tablet replicas per shard per table. Suppressed by tablet hints in table's schema: min_per_shard_tablet_count and min_tablet_count")
+    , tablets_per_shard_goal(this, "tablets_per_shard_goal", liveness::LiveUpdate, value_status::Used, 100,
+         "The goal for the maximum number of tablet replicas per shard. Tablet allocator tries to keep this goal.")
```
nit in commit tablets, topology: Introduce tablets_per_shard_goal config log message: s/decrese/decrease/, s/similat/similar/

bhalevy commented on db/config.cc:
```diff
@@ -1372,6 +1373,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , _extensions(std::move(exts))
 {
     add_tombstone_gc_extension();
+    _extensions->add_schema_extension<db::tablet_hints_extension>(db::tablet_hints_extension::NAME);
```
This patch is not needed after the options were promoted to scylla_tables (from schema extensions)

bhalevy commented on locator/network_topology_strategy.cc:
```diff
@@ -350,14 +350,7 @@ future<size_t> network_topology_strategy::calculate_min_tablet_count(schema_ptr
-        tablet_count = aligned_tablet_count;
-    }
-
+future<tablet_map> network_topology_strategy::allocate_tablets_for_new_table(schema_ptr s, token_metadata_ptr tm, size_t tablet_count) const {
```
The patch needs more justification to explain the motivation and why is it correct.

bhalevy commented on locator/network_topology_strategy.hh:
```diff
@@ -29,7 +29,7 @@ public:
     }
 
-    size_t get_replication_factor(const sstring& dc) const {
+    size_t get_replication_factor(const sstring& dc) const override {
```
nit in git log message: s/controling/controlling/

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+
+        std::unordered_map<sstring, unsigned> shards_per_dc;
+        _tm->get_topology().for_each_node([&] (const node& n) {
+            if (n.is_normal() && _tm->is_normal_token_owner(n.host_id())) {
```
for_each_node + is_normal_token_owner = for_each_token_owner

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+
+            for (auto&& [table, table_plan] : plan.tables) {
+                auto* rs = rs_by_table[table];
+                if (!rs) {
```
Under what conditions does that happen?
Worth a comment...

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+
+            bool overloaded = avg_tablets_per_shard > _tablets_per_shard_goal;
+            lblogger.debug("new_avg_tablets_per_shard[dc={}]: {:.3f}{}", dc, avg_tablets_per_shard,
+                overloaded ? " (overloaded!)" : "");
```
It would be better if the variables followed the same naming convention as the log messages.
It'd be confusing otherwise.

bhalevy replied:
maybe use `cur_` and `new_` to make it clearer which is which

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+                lblogger.debug("old_avg_tablets_per_shard [dc={}, table={}]: {:.3f}", dc, table, old_tablets_per_shard);
+
+                auto tablets_per_shard = get_avg_tablets_per_shard(table_plan.target_tablet_count);
+                avg_tablets_per_shard += tablets_per_shard;
```
I don't see where that's normalized by the number of tables.
Or did I miss something?

tgrabiec replied:
Why should we normalize by the number of tables? We're computing per-shard load, because the limit is for tablets per shard. Number of tables doesn't change anything. We can have one table with 100 tablets per shard, or 100 tables with 1 tablet per shard.

bhalevy replied:
ok, got it.  It is the per-shard tablet count goal we're interested in here, not per table per_shard_tablet_count.

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+                auto scale = _tablets_per_shard_goal / avg_tablets_per_shard;
+
+                for (auto&& [table, table_plan]: plan.tables) {
+                    auto *rs = rs_by_table[table];
```
nit: `*` should stick to the left

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+
+                for (auto&& [table, table_plan]: plan.tables) {
+                    auto *rs = rs_by_table[table];
+                    if (!rs) {
```
again, how can that happen?

bhalevy replied:
Just in test_mode?

tgrabiec replied:
Yes, in test_mode (which is going away).

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+                    // in other DCs.
+                    if (rf) {
+                        if (table_scaling.contains(table)) {
+                            table_scaling[table] = std::min(table_scaling[table], scale);
```
Instead of search `table` twice, how about using `find` to get a read/write iterator?

bhalevy replied:
Can even use try_emplace(table, scale) and use the resulting iterator for calculating and storing the minimum (if needed) if the entry was not inserted.

tgrabiec replied:
Done

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1127,123 @@ class load_balancer {
+            table_plan.target_tablet_count_aligned = 1u << log2ceil(table_plan.target_tablet_count);
+
+            if (table_plan.target_tablet_count_aligned > table_plan.current_tablet_count) {
+                table_plan.resize_decision = locator::resize_decision::split();
```
This patch is about scaling down, can this case actually happen at this stage?

tgrabiec replied:
This computes the resize decision after every adjustments, not just scaling. The block was moved here.

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1059,20 +1059,70 @@ class load_balancer {
+                if (min_per_shard_tablet_count) {
+                    target_tablet_count = std::max<size_t>(target_tablet_count,
+                            tablet_count_from_min_per_shard_tablet_count(*s, shards_per_dc, *rs, min_per_shard_tablet_count));
+                }
```
Will review after rebase (if this is still needed)

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1084,6 +1084,35 @@ class load_balancer {
         return tablet_count;
     }
 
+    std::tuple<schema_ptr, const tablet_aware_replication_strategy*> get_schema_and_rs(table_id table) {
```
nit: in git log s/readibility/readability/

Also, it'd probably be better to move this patch earlier in the series before any functional changes are made.

bhalevy commented on test/boost/tablets_test.cc:
```diff
@@ -1487,7 +1487,7 @@ void do_rebalance_tablets(cql_test_env& e,
 
     for (size_t i = 0; i < max_iterations; ++i) {
-        auto plan = talloc.balance_tablets(stm.get(), load_stats, skiplist, true).get();
+        auto plan = talloc.balance_tablets(stm.get(), load_stats, skiplist).get();
```
This patch looks good.
It could be done as part of the test change by the way, and can be committed independently if needed.

tgrabiec replied:
Yeah, but the test change was independent of tablet options PR which introduced test_mode.

bhalevy commented on test/topology_custom/test_tablets.py:
```diff
@@ -45,6 +45,21 @@ async def test_tablet_replication_factor_enough_nodes(manager: ManagerClient):
+
+    cql = manager.get_cql()
+
+    await cql.run_async(f"CREATE KEYSPACE test WITH replication = {{'class': 'NetworkTopologyStrategy', 'replication_factor': 1}}")
```
If this PR lands after https://github.com/scylladb/scylladb/pull/22399, please use `async with new_test_keyspace`.

bhalevy commented on locator/network_topology_strategy.cc:
```diff
@@ -336,12 +336,16 @@ future<size_t> network_topology_strategy::calculate_min_tablet_count(schema_ptr
+            (get_initial_tablets() || tablet_options.min_tablet_count) ? 0 : initial_scale.value_or(1));
+    if (min_per_shard_tablet_count) {
+        tablet_count = std::max<size_t>(tablet_count, co_await calculate_initial_tablets_from_topology(*s, tm, _dc_rep_factor,
+                                                              min_per_shard_tablet_count));
```
After this change there should not be any call sites left with the default value for `min_per_shard_tablet_count`, so we can change the function signature to require it in all cases.

bhalevy replied:
And also drop `if (min_per_shard_tablet_count)` there,  We can even `SCYLLA_ASSERT(min_per_shard_tablet_count > 0.)` instead.

tgrabiec replied:
> And also drop `if (min_per_shard_tablet_count)` there, We can even `SCYLLA_ASSERT(min_per_shard_tablet_count > 0.)` instead.

min_per_shard_tablet_count can still be 0 if `get_initial_tablets() || tablet_options.min_tablet_count`

bhalevy replied:
> > And also drop `if (min_per_shard_tablet_count)` there, We can even `SCYLLA_ASSERT(min_per_shard_tablet_count > 0.)` instead.
> 
> min_per_shard_tablet_count can still be 0 if `get_initial_tablets() || tablet_options.min_tablet_count`

But then we don't call calculate_initial_tablets_from_topology

tgrabiec replied:
> > > And also drop `if (min_per_shard_tablet_count)` there, We can even `SCYLLA_ASSERT(min_per_shard_tablet_count > 0.)` instead.
> > 
> > 
> > min_per_shard_tablet_count can still be 0 if `get_initial_tablets() || tablet_options.min_tablet_count`
> 
> But then we don't call calculate_initial_tablets_from_topology

Right, I misunderstood. Well, that if is removed by later patches, so I don't think there's anything to do.

bhalevy replied:
okay.  Idealy, there was room for leaving FIXME comments documenting the intentions for that condition and also the std::max for the min per-shard tablet count that goes away in a later patch, to make it easier on the reviewers, but let's not hokd up this PR on that.

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1076,45 +1054,122 @@ class load_balancer {
+    struct table_sizing {
+        size_t current_tablet_count; // Tablet count in group0.
+        size_t target_tablet_count; // Tablet count wanted by scheduler.
+        std::optional<uint64_t> avg_tablet_size; // nullptr when stats not yet available.
```
nit: s/nullptr/nullopt/

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1076,45 +1054,122 @@ class load_balancer {
+                if (avg_tablet_size > target_max_tablet_size() ||
+                    (cur_decision.is_split() && avg_tablet_size >= _target_tablet_size)) {
+                    // TODO: extend to n-way split when needed
+                    tablet_count_from_size *= 2;
```
This does much more than refactoring the existing code in to_resize_decision.
Is there a way to break this patch into two: first one does only refactoring and the following patch implementing the changes to the existing algorithm?

bhalevy replied:
> This does much more than refactoring the existing code in to_resize_decision, as the patch title suggests.
Is there a way to break this patch into two: first one does only refactoring and the following patch implementing the changes to the existing algorithm?

tgrabiec replied:
I can do that.

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1128,6 +1128,10 @@ class load_balancer {
 
+        if (new_table) {
+            co_await process_table(new_table->id(), new_table, new_rs, 0);
+        }
```
nit: in patch git log message: s/decisisons/decisions/

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1132,6 +1131,114 @@ class load_balancer {
+
+            {
+                bool overloaded = cur_avg_tablets_per_shard > _tablets_per_shard_goal;
+                lblogger.info("new_avg_tablets_per_shard[dc={}]: {:.3f}{}", dc, cur_avg_tablets_per_shard,
```
s/new_/cur_/ in format string.

tgrabiec replied:
Ouch

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1132,6 +1131,114 @@ class load_balancer {
+            lblogger.debug("Table {}, {} => {} ({}), resize: {}", table,
+                           table_plan.current_tablet_count,
+                           table_plan.target_tablet_count_aligned,
+                           table_plan.target_tablet_count,
```
Switching between `target_tablet_count_aligned` and `target_tablet_count` makes more sense IMO, as after split or merge `current_tablet_count` should end up being `target_tablet_count_aligned`, no?
so it should be to the right of `=>`.

tgrabiec replied:
> Switching between `target_tablet_count_aligned` and `target_tablet_count` makes more sense IMO, as after split or merge `current_tablet_count` should end up being `target_tablet_count_aligned`, no? so it should be to the right of `=>`.

Current pattern is:

`current_tablet_count => target_tablet_count_aligned (target_tablet_count)`

what do you propose?

bhalevy replied:
Oh, I missed the `table` parameter, that goes before the first comma...

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1072,19 +1072,69 @@ class load_balancer {
+            double min_per_shard_tablet_count)
+    {
+        // Try to use as many tablets so that all shards in the current topology
+        // are covered with at least `min_per_shard_tablet_count` tablets on avergae.
```
nit: s/avergae/average/

bhalevy commented on service/tablet_allocator.cc:
```diff
@@ -1076,45 +1054,122 @@ class load_balancer {
+
+        auto process_table = [&] (table_id table, schema_ptr s, const tablet_aware_replication_strategy* rs, size_t tablet_count) -> future<> {
+            table_sizing& table_plan = plan.tables[table];
+            table_plan.current_tablet_count = tablet_count;
```
Also, I see that "tablets: load_balancer: Move hints processing to tablet scheduler" makes this path synchronous again and adds coroutine::maybe_yield().
Is there any sense in coroutinizing it in the early patches in the series then?

tgrabiec replied:
That would duplicate the logic when process_table() is called from multiple places.

tgrabiec replied:
> Also, I see that "tablets: load_balancer: Move hints processing to tablet scheduler" makes this path synchronous again and adds coroutine::maybe_yield(). Is there any sense in coroutinizing it in the early patches in the series then?

Probably not much, except for better bisectability. If I didn't make this a coroutine I could get comments on earlier patches that I'm introducing stalls.

bhalevy commented on test/boost/tablets_test.cc:
```diff
@@ -2615,6 +2615,73 @@ SEASTAR_THREAD_TEST_CASE(test_per_shard_goal_mixed_dc_rf) {
+
+        auto ks_name1 = add_keyspace(e, {{dc, 1}}, 2);
+        e.execute_cql(fmt::format("CREATE TABLE {}.table1 (p1 text, r1 int, PRIMARY KEY (p1)) "
+                                  "WITH tablets = {{}}", ks_name1)).get();
```
why is `WITH tablets = {{}}` needed?

tgrabiec replied:
It's not, just copy pasta. I'll remove it.

bhalevy commented on test/boost/tablets_test.cc:
```diff
@@ -2615,6 +2615,73 @@ SEASTAR_THREAD_TEST_CASE(test_per_shard_goal_mixed_dc_rf) {
+        // merge
+        cfg.tablets_per_shard_goal(100);
+        rebalance_tablets(e, &load_stats);
+        BOOST_REQUIRE_EQUAL(get_tablet_count(), 16);
```
nice test! thanks!

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15204/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15204/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/network_topology_strategy_test
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_major_compaction
:small_blue_diamond: topology_custom/test_tablets
#### Failed Tests (60/1262):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.1.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.1.xunit/xml/_empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.1.xunit/xml/_empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.10.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.10.xunit/xml/_empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.10.xunit/xml/_empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.2.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.2.xunit/xml/_empty__2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.2.xunit/xml/_empty__3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15204/testReport/junit/boost.network_topology_strategy_test.NetworkTopologyStrategy_tablet_allocation_balancing_test.3.xunit/xml/_empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 3 hr 25 min
- Builder: i-00bbad20592c51c5a (m5ad.8xlarge)

tgrabiec commented on service/tablet_allocator.cc:
```diff
@@ -1132,6 +1131,114 @@ class load_balancer {
+        for (auto&& [table, scale] : table_scaling) {
+            auto& table_plan = plan.tables[table];
+            auto new_count = std::max<size_t>(1, table_plan.target_tablet_count * scale);
+            lblogger.debug("Scaling down table {} by a factor of {:.3f}: {} => {}", table, scale, table_plan.target_tablet_count, new_count);
```
But it's debug level, and knowing that scaling was not effective for a given table is some information too.

tgrabiec commented:
In V3:
  - rebased on top of master, which now has the tablet options PR in, and the topology_builder in tests PR
  - added more unit tests
  - dropped test_mode
  - use for_each_token_owner() 
  - make "initial" tablet option on keyspace level inhibit initial_scale (regression fix)
  - moved get_schema_rs() refactoring earlier
  - made tablet config options which control tablet count live-updateable
  - made tablet_per_shard_goal a scylla.yaml config instead of group0 state so that it fits with current backup and setup procedures
  - log reason for tablet count during load balancing
  - fixed typos

tgrabiec commented:
In V4:
  - Split the path which extracts make_sizing_plan() into multiple patches for easier review and clearer intent 
  - Fixed typos, white space
  - fixed failing network_topology_strategy_test, due to missing alignment of tablet count
  - nullptr -> nullopt in comment above avg_tablet_size
  - improved bisectability
  - dropped unnecessary "with tablets = {}" in CQL query in tests

tgrabiec commented:
In V4.1:
  - fixed even more typos

```range-diff
 1:  1adf0602d5 !  1:  719ef2f7e7 tablets: load_balancer: Determine desried count from size separately from count from options
    @@ Metadata
     Author: Tomasz Grabiec <tgrabiec@scylladb.com>
     
      ## Commit message ##
    -    tablets: load_balancer: Determine desried count from size separately from count from options
    +    tablets: load_balancer: Determine desired count from size separately from count from options
     
    -    For debugging purposes, later we will want to know which rule
    +    For debugging purposes. Later we will want to know which rule
         determined the count.
     
      ## service/tablet_allocator.cc ##
 2:  da1418f780 =  2:  3e66bc99f2 tablets: Use scheduler's make_sizing_plan() to decide about tablet count of a new table
 3:  06fe27ccc4 !  3:  4e9bfe1d7a tablets: load_balancer: Scale down tablet count to respect per-shard tablet count goal
    @@ Commit message
         still ok. It controls tablet allocator behavior, when choosing initial
         tablet count. Even though it's a per-node config, we don't support
         different limits per node. All nodes must have the same value of that
    -    config. It's similat in that regard to other scheduler config items
    +    config. It's similar in that regard to other scheduler config items
         like tablets_initial_scale_factor and target_tablet_size_in_bytes.
     
      ## db/config.cc ##
    @@ locator/tablet_replication_strategy.hh: class tablet_aware_replication_strategy
          /// Runs under group0 guard.
          virtual future<tablet_map> reallocate_tablets(schema_ptr, token_metadata_ptr, tablet_map cur_tablets) const = 0;
     +
    -+    /// Returns desired replication factor in a given DC.
    ++    /// Returns replication factor in a given DC.
     +    /// Note that individual tablets may lag behind desired replication factor in their
    -+    /// current replica list, as replication factor changes invole table rebuilding transitions
    -+    /// which are not instantenous.
    ++    /// current replica list, as replication factor changes involve table rebuilding transitions
    ++    /// which are not instantaneous.
     +    virtual size_t get_replication_factor(const sstring& dc) const = 0;
      };
```

bhalevy reviewed (approved):
LGTM

raphaelsc commented:
@tgrabiec without diving into the code, if we multiple tables are created in a quick succession, leading avg tablet count in a DC to exceed the goal, do we use resize urgency to determine which table to resize first? do we make sure we just resize as many tables as needed to reach the goal?

tgrabiec commented:
> @tgrabiec without diving into the code, if we multiple tables are created in a quick succession, leading avg tablet count in a DC to exceed the goal, do we use resize urgency to determine which table to resize first? do we make sure we just resize as many tables as needed to reach the goal?

There is a mitigation factor - new tables will be allocated fewer and fewer tablets, according to scaling applied by the scheduler as if the table was already created. All tables are scaled by the same factor.

Resize scheduling was not changed. This should be a follow-up, because resizing one table at a time may limit merge speed.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15222/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15222/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/network_topology_strategy_test
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_major_compaction
:small_blue_diamond: topology_custom/test_tablets
#### Failed Tests (1/3612):
- [test_replace_with_no_normal_token_owners_in_dc[False].release.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15222/testReport/junit/topology_custom/test_tablets/test_replace_with_no_normal_token_owners_in_dc_False__release_10) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc[False].release.10)
#### Build Details:
- Duration: 5 hr 42 min
- Builder: i-066654c62e45d31f5 (m5ad.8xlarge)

tgrabiec commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15222/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15222/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/network_topology_strategy_test 🔹 boost/tablets_test 🔹 topology_custom/test_major_compaction 🔹 topology_custom/test_tablets
> 
> #### Failed Tests (1/3612):
> * [test_replace_with_no_normal_token_owners_in_dc[False].release.10](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15222/testReport/junit/topology_custom/test_tablets/test_replace_with_no_normal_token_owners_in_dc_False__release_10) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_no_normal_token_owners_in_dc%5BFalse%5D.release.10)
> 
> #### Build Details:
> * Duration: 5 hr 42 min
> * Builder: i-066654c62e45d31f5 (m5ad.8xlarge)

https://github.com/scylladb/scylladb/issues/22774

mykaul commented:
@avikivity - please re-review.

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15250/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15250/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/network_topology_strategy_test
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_major_compaction
:small_blue_diamond: topology_custom/test_tablets
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14542/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6523/)
:x: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3406/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15250/)
#### Failed Tests (14/43838):
- [test_queries_during_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/reversed_queries_test/TestReversedQueriesReadRepairDuringUpgrade/Tests___dtest_with_topology_changes___test_queries_during_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_queries_during_upgrade)
- [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/rolling_upgrade_test/TestRollingUpgrade/Tests___dtest_with_topology_changes___test_rolling_upgrade) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
- [test_supported_cipher_algorithms[local]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_local_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms[local])
- [test_supported_cipher_algorithms[replicated]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_replicated_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms[replicated])
- [test_supported_cipher_algorithms[kmip]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_kmip_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms[kmip])
- [test_supported_cipher_algorithms[kms]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_kms_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms[kms])
- [test_supported_cipher_algorithms[kms_real]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_kms_real_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms[kms_real])
- [test_ics_with_partitions_larger_than_sstable_size](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/ics_compaction_test/TestIcsCompaction/Tests___dtest_with_tablets___test_ics_with_partitions_larger_than_sstable_size) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ics_with_partitions_larger_than_sstable_size)
- [test_ics_snapshot_and_restore](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/ics_compaction_test/TestIcsCompaction/Tests___dtest_with_tablets___test_ics_snapshot_and_restore) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ics_snapshot_and_restore)
- [test_space_amplification_goal_trigger](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/ics_compaction_test/TestIcsCompaction/Tests___dtest_with_tablets___test_space_amplification_goal_trigger) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_space_amplification_goal_trigger)
#### Build Details:
- Duration: 20 hr
- Builder: i-04d4ce2afed6a5914 (m5ad.8xlarge)

bhalevy commented:
@tgrabiec please analyze CI falires.
E.g. https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6523/artifact/logs-full.release.007/1739435032090_rolling_upgrade_test.py%3A%3ATestRollingUpgrade%3A%3Atest_rolling_upgrade/node2.log
```
ERROR 2025-02-13 08:23:15,622 [shard 0: gms] topology_state_machine - cannot map name tablet resize finalization to transition_state, at: 0x61831de 0x61837f0 0x6183af8 0x5c47d07 0x4871d86 0x338522f 0x33474ca 0x5c85dbf 0x5c8733a 0x5c88517 0x5c878c8 0x5c17a33 0x5c16d93 0x13cf425 0x13d0de0 0x13cd843 /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.ccm/scylla-repository/release/6.2.3/libreloc/libc.so.6+0x2a087 /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.ccm/scylla-repository/release/6.2.3/libreloc/libc.so.6+0x2a14a 0x13caec4
```

raphaelsc commented:
#22840

On Sun, Feb 16, 2025, 8:29 AM Benny Halevy ***@***.***> wrote:

> @tgrabiec <https://github.com/tgrabiec> please analyze CI falires.
> E.g.
> https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6523/artifact/logs-full.release.007/1739435032090_rolling_upgrade_test.py%3A%3ATestRollingUpgrade%3A%3Atest_rolling_upgrade/node2.log
>
> ERROR 2025-02-13 08:23:15,622 [shard 0: gms] topology_state_machine - cannot map name tablet resize finalization to transition_state, at: 0x61831de 0x61837f0 0x6183af8 0x5c47d07 0x4871d86 0x338522f 0x33474ca 0x5c85dbf 0x5c8733a 0x5c88517 0x5c878c8 0x5c17a33 0x5c16d93 0x13cf425 0x13d0de0 0x13cd843 /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.ccm/scylla-repository/release/6.2.3/libreloc/libc.so.6+0x2a087 /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.ccm/scylla-repository/release/6.2.3/libreloc/libc.so.6+0x2a14a 0x13caec4
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/22522#issuecomment-2661387070>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKYA43PMPCI2R6U4U4H32T2QBZBJAVCNFSM6AAAAABV6SKGSWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGM4DOMBXGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
> [image: bhalevy]*bhalevy* left a comment (scylladb/scylladb#22522)
> <https://github.com/scylladb/scylladb/pull/22522#issuecomment-2661387070>
>
> @tgrabiec <https://github.com/tgrabiec> please analyze CI falires.
> E.g.
> https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6523/artifact/logs-full.release.007/1739435032090_rolling_upgrade_test.py%3A%3ATestRollingUpgrade%3A%3Atest_rolling_upgrade/node2.log
>
> ERROR 2025-02-13 08:23:15,622 [shard 0: gms] topology_state_machine - cannot map name tablet resize finalization to transition_state, at: 0x61831de 0x61837f0 0x6183af8 0x5c47d07 0x4871d86 0x338522f 0x33474ca 0x5c85dbf 0x5c8733a 0x5c88517 0x5c878c8 0x5c17a33 0x5c16d93 0x13cf425 0x13d0de0 0x13cd843 /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.ccm/scylla-repository/release/6.2.3/libreloc/libc.so.6+0x2a087 /jenkins/workspace/scylla-master/gating-dtest-release-with-consistent-topology-changes/scylla/.ccm/scylla-repository/release/6.2.3/libreloc/libc.so.6+0x2a14a 0x13caec4
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/22522#issuecomment-2661387070>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKYA43PMPCI2R6U4U4H32T2QBZBJAVCNFSM6AAAAABV6SKGSWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRRGM4DOMBXGA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

tgrabiec commented:
> #### 🔴 CI State: ABORTED
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15250/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15250/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/network_topology_strategy_test 🔹 boost/tablets_test 🔹 topology_custom/test_major_compaction 🔹 topology_custom/test_tablets ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14542/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6523/) ❌ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3406/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15250/)
> 
> #### Failed Tests (14/43838):
> * [test_queries_during_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/reversed_queries_test/TestReversedQueriesReadRepairDuringUpgrade/Tests___dtest_with_topology_changes___test_queries_during_upgrade) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_queries_during_upgrade)
> * [test_rolling_upgrade](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/rolling_upgrade_test/TestRollingUpgrade/Tests___dtest_with_topology_changes___test_rolling_upgrade) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade)
> * [test_supported_cipher_algorithms[local]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_local_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms%5Blocal%5D)
> * [test_supported_cipher_algorithms[replicated]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_replicated_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms%5Breplicated%5D)
> * [test_supported_cipher_algorithms[kmip]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_kmip_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms%5Bkmip%5D)
> * [test_supported_cipher_algorithms[kms]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_kms_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms%5Bkms%5D)
> * [test_supported_cipher_algorithms[kms_real]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/encryption_at_rest_test/TestEncryptionAtRest/Tests___dtest_with_tablets___test_supported_cipher_algorithms_kms_real_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_supported_cipher_algorithms%5Bkms_real%5D)
> * [test_ics_with_partitions_larger_than_sstable_size](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/ics_compaction_test/TestIcsCompaction/Tests___dtest_with_tablets___test_ics_with_partitions_larger_than_sstable_size) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ics_with_partitions_larger_than_sstable_size)
> * [test_ics_snapshot_and_restore](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/ics_compaction_test/TestIcsCompaction/Tests___dtest_with_tablets___test_ics_snapshot_and_restore) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_ics_snapshot_and_restore)
> * [test_space_amplification_goal_trigger](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/15250/testReport/junit/ics_compaction_test/TestIcsCompaction/Tests___dtest_with_tablets___test_space_amplification_goal_trigger) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_space_amplification_goal_trigger)
> 
> #### Build Details:
> * Duration: 20 hr
> * Builder: i-04d4ce2afed6a5914 (m5ad.8xlarge)

Some failures are due to: https://github.com/scylladb/scylladb/issues/21859

bhalevy commented:
@tgrabiec dtest fix is in.
Please rebase and resolve conflicts.

tgrabiec commented:
> @tgrabiec dtest fix is in. Please rebase and resolve conflicts.

https://github.com/scylladb/scylladb/issues/21859 is not fixed yet though.

raphaelsc commented:
> > > @tgrabiec dtest fix is in. Please rebase and resolve conflicts.
> > 
> > 
> > #21859 is not fixed yet though.
> 
> Submitting the fix in less than 1 hour.

https://github.com/scylladb/scylladb/pull/22933

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15450/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15450/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/network_topology_strategy_test
:small_blue_diamond: boost/tablets_test
:small_blue_diamond: topology_custom/test_major_compaction
:small_blue_diamond: topology_custom/test_tablets
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/3558/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/6680/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/14698/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15450/)
#### Build Details:
- Duration: 9 hr 18 min
- Builder: i-00c92b183cf242102 (m5ad.12xlarge)

tgrabiec commented:
In V4:
  - rebased
  - changed avg-per-shard-load-per-dc logging to debug() to match with recent Avi's change to silence the load balancer
  - use initial scale of 2 (debug/dev) and 4 (release) in all test.py tests to avoid timeouts and slow down

range-diff (the interesting parts):
```
  4:  f317cd33d7 ! 106:  2463e524ed tablets: Set default initial tablet count scale to 10
    @@ Commit message
     
         Fixes #21967
     
    +    In some tests, we explicity set the initial scale to 1 as some of the
    +    existing tests assume 1 compaction group per shard.
    +
    +    test.py uses a lower default. Having many tablets per shard slows down
    +    certain topology operations like decommission/replace/removenode,
    +    where the running time is proportional to tablet count, not data size,
    +    because constant cost (latency) of migration dominates. This latency
    +    is due to group0 operations and barriers. This is especially
    +    pronounced in debug mode. Scheduler allows at most 2 migrations per
    +    shard, so this latency becomes a determining factor for decommission
    +    speed.
    +
    +    To avoid this problem in tests, we use lower default for tablet count per
    +    shard, 2 in debug/dev mode and 4 in release mode. Alternatively, we
    +    could compensate by allowing more concurrency when migrating small
    +    tablets, but there's no infrastructure for that yet.
    +
    +    I observed that with 10 tablets per shard, debug-mode
    +    topology_custom.mv/test_mv_topology_change starts to time-out during
    +    removenode (30 s).
    +
      ## db/config.cc ##
     @@ db/config.cc: db::config::config(std::shared_ptr<db::extensions> exts)
          , minimum_replication_factor_warn_threshold(this, "minimum_replication_factor_warn_threshold", liveness::LiveUpdate, value_status::Used,  3, "")
    @@ db/config.cc: db::config::config(std::shared_ptr<db::extensions> exts)
               "Allows target tablet size to be configured. Defaults to 5G (in bytes). Maintaining tablets at reasonable sizes is important to be able to " \
               "redistribute load. A higher value means tablet migration throughput can be reduced. A lower value may cause number of tablets to increase significantly, " \
     
    + ## test.py ##
    +@@ test.py: class PythonTestSuite(TestSuite):
    +             default_config_options = \
    +                     {"authenticator": "PasswordAuthenticator",
    +                      "authorizer": "CassandraAuthorizer"}
    ++            default_config_options["tablets_initial_scale_factor"] = 4 if self.mode == "release" else 2
    +             config_options = default_config_options | \
    +                              self.cfg.get("extra_scylla_config_options", {}) | \
    +                              create_cfg.config_from_test
    +
      ## test/cqlpy/suite.yaml ##
     @@ test/cqlpy/suite.yaml: extra_scylla_cmdline_options:
        - '--experimental-features=keyspace-storage-options'
```

tgrabiec commented:
> #### 🟢 CI State: SUCCESS

I think the change to dtest to use initial scale of 2 made https://github.com/scylladb/scylladb/issues/21859 less likely again.

avikivity commented:
Queued

bhalevy commented:
@tgrabiec what's the backport status of this PR?
According to https://github.com/scylladb/scylladb/issues/23284#issuecomment-2733362055 I understand we do want it.
Are there any prerequisites for backport?

tgrabiec commented:
> @tgrabiec what's the backport status of this PR? According to [#23284 (comment)](https://github.com/scylladb/scylladb/issues/23284#issuecomment-2733362055) I understand we do want it. 

Not backported yet. We didn't want it in 2025.1.0, so it was not backported to branch-2025.1. @mykaul explained that there is some process which will ensure we won't miss it, so I'm just waiting for signal.  

> Are there any prerequisites for backport?

Yes, per-table tablet options is a prerequistie.

mykaul commented:
> > @tgrabiec what's the backport status of this PR? According to [#23284 (comment)](https://github.com/scylladb/scylladb/issues/23284#issuecomment-2733362055) I understand we do want it.
> 
> Not backported yet. We didn't want it in 2025.1.0, so it was not backported to branch-2025.1. @mykaul explained that there is some process which will ensure we won't miss it, so I'm just waiting for signal.

We (I) wanted to wait for some test results on master to see that we are happy with it. If we are, let's backport.
